File: c_style_check/go/c_style_check.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strings"
)

const MAX_LINE_LENGTH = 80

var tabs = regexp.MustCompile(`\t+`)
var comma_space = regexp.MustCompile(`,[^ ]`)
var operator_space = regexp.MustCompile(`(\w(\+|\-|\*|\<|\>|\=)\w)|(\w(\=\=|\<\=|\>\=)\w)`)
var comment_line = regexp.MustCompile(`^\s*\/\*.*\*\/\s*$`)
var open_comment_space = regexp.MustCompile(`\/\*[^ *\n]`)
var close_comment_space = regexp.MustCompile(`[^ *]\*\/`)
var paren_curly_space = regexp.MustCompile(`\)\{`)
var c_plus_plus_comment = regexp.MustCompile(`\/\/`)
var semi_space = regexp.MustCompile(`;[^ \s]`)

func checkLine(filename string, line string, n int) {
	// Strip the terminal newline.
	line = strings.TrimRight(line, "\n")

	if tabs.MatchString(line) {
		fmt.Printf("File: %s, line %d: [TABS]:\n%s", filename, n, line)
	}

	if len(line) > MAX_LINE_LENGTH {
		fmt.Printf("File: %s, line %d: [TOO LONG (%d CHARS)]:\n%s\n", filename, n, len(line), line)
	}

	if open_comment_space.MatchString(line) {
		fmt.Printf("File: %s, line: %d: [PUT SPACE AFTER OPEN COMMENT]:\n%s\n", filename, n, line)
	}

	if close_comment_space.MatchString(line) {
		fmt.Printf("File: %s, line: %d: [PUT SPACE AFTER CLOSE COMMENT]:\n%s\n", filename, n, line)
	}

	if paren_curly_space.MatchString(line) {
		fmt.Printf("File: %s, line: %d: [PUT SPACE BETWEEN ) AND {]:\n%s\n", filename, n, line)
	}

	if c_plus_plus_comment.MatchString(line) {
		fmt.Printf("File: %s, line: %d: [DON'T USE C++ COMMENTS]:\n%s\n", filename, n, line)
	}

	if semi_space.MatchString(line) {
		fmt.Printf("File: %s, line: %d: [PUT SPACE/NEWLINE AFTER SEMICOLON]:\n%s\n", filename, n, line)
	}

	if operator_space.MatchString(line) {
		if !comment_line.MatchString(line) {
			fmt.Printf("File: %s, line: %d: [PUT SPACE AROUND OPERATORS]:\n%s\n", filename, n, line)
		}
	}

	if comma_space.MatchString(line) {
		fmt.Printf("File: %s, line: %d: [PUT SPACE AFTER COMMA]:\n%s\n", filename, n, line)
	}
}

func checkFile(filename string) {
	file, err := os.Open(filename)
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	defer file.Close()

	i := 0
	// Read file line by line.
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		i++ // Start on line 1.
		line := scanner.Text()
		checkLine(filename, line, i)
	}
}

func main() {
	usage := `usage: c_style_check filename1 [filename2 ...]`

	if len(os.Args) < 2 {
		fmt.Println(usage)
		os.Exit(0)
	}

	for _, filename := range os.Args[1:] {
		checkFile(filename)
	}
}

File: c_style_check/julia/c_style_check.jl
const MAX_LINE_LENGTH = 80

const tabs = r"\t+"
const comma_space = r",[^ ]"
const operator_space = r"(\w(\+|\-|\*|\<|\>|\=)\w)|(\w(\=\=|\<\=|\>\=)\w)"
const comment_line = r"^\s*\/\*.*\*\/\s*$"
const open_comment_space = r"\/\*[^ *\n]"
const close_comment_space = r"[^ *]\*\/"
const paren_curly_space = r"\)\{"
const c_plus_plus_comment = r"//"
const semi_space = r";[^ \s]"

function check_line(filename::String, line::String, n::Int)
    # Strip the terminal newline
    line = rstrip(line)

    if occursin(tabs, line)
        print("File: $(filename), line: $(n): [TABS]:\n$(line)\n")
    end

    if length(line) > MAX_LINE_LENGTH
        print("File: $(filename), line: $(n): [TOO LONG (88 CHARS)]:\n$(line)\n")
    end

    if occursin(open_comment_space, line)
        print("File: $(filename), line: $(n): [PUT SPACE AFTER OPEN COMMENT]:\n$(line)\n")
    end

    if occursin(close_comment_space, line)
        print("File: $(filename), line: $(n): [PUT SPACE AFTER CLOSE COMMENT]:\n$(line)\n")
    end

    if occursin(paren_curly_space, line)
        print("File: $(filename), line: $(n): [PUT SPACE BETWEEN ) AND {]:\n$(line)\n")
    end

    if occursin(c_plus_plus_comment, line)
        print("File: $(filename), line: $(n): [DON'T USE C++ COMMENTS]:\n$(line)\n")
    end

    if occursin(semi_space, line)
        print(
            "File: $(filename), line: $(n): [PUT SPACE/NEWLINE AFTER SEMICOLON]:\n$(line)\n",
        )
    end

    if occursin(operator_space, line)
        if !occursin(comment_line, line)
            print("File: $(filename), line: $(n): [PUT SPACE AROUND OPERATORS]:\n$(line)\n")
        end
    end

    if occursin(comma_space, line)
        print("File: $(filename), line: $(n): [PUT SPACE AFTER COMMA]:\n$(line)\n")
    end
end

function check_file(filename::String)
    i = 0
    open(filename) do file
        for line in eachline(file)
            i += 1
            check_line(filename, line, i)
        end
    end
end

function usage()
    println("usage: c_style_check filename1 [filename2 ...]")
    return
end

function main(args::Vector{String})
    if length(args) < 1
        usage()
    end

    for arg in ARGS
        check_file(arg)
    end
end

main(ARGS)
